File: Python/_105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        return self.helper(preorder, 0, len(preorder) - 1, inorder, 0, len(inorder) - 1)
        
    def helper(self, preorder, preStart, preEnd, inorder, inStart, inEnd):
        if inStart > inEnd or preStart > preEnd:
            return None
        rootVal = preorder[preStart]
        rootIndex = 0
        for i in range(inStart, inEnd + 1):
            if inorder[i] == rootVal:
                rootIndex = i
                break
        length = rootIndex - inStart
        root = TreeNode(rootVal)
        root.left = self.helper(preorder, preStart + 1, preStart + length, inorder, inStart, rootIndex - 1)
        root.right = self.helper(preorder, preStart + length + 1, preEnd, inorder, rootIndex + 1, inEnd)    
        return root

File: Java/_105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }
    
    public TreeNode buildTree(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd) {
        if (inStart > inEnd || preStart > preEnd)
            return null;
        int rootVal = preorder[preStart];
        int rootIndex = 0;
        for (int i = inStart; i <= inEnd; i++) {
            if (inorder[i] == rootVal) {
                rootIndex = i;
                break;
            }
        }
        int len = rootIndex - inStart;
        TreeNode root = new TreeNode(rootVal);
        root.left = buildTree(preorder, preStart + 1, preStart + len, inorder, inStart, rootIndex - 1);
        root.right = buildTree(preorder, preStart + len + 1, preEnd, inorder, rootIndex + 1, inEnd);
        
        return root;
    }
}
